### PR TITLE
EVG-18784 Warn users when creating/copying projects at limit

### PR DESCRIPTION
**Pull request by bynn**
[EVG-18784](https://jira.mongodb.org/browse/EVG-18784)

### Description 
when creating/copying projects on spruce, users will now receive a warning toast 

the validation function has been updated to return an error code and handle all edge cases that it didn't handle before 
(test has been added for that)

### Testing 
staging
creating and copying projects that are at the limit will warn the users
    
#### Does this need documentation?
eventually (there is a ticket for create project documentation)

### CODE REVIEW

**Comment by bynn** [comment]
evergreen retry

File: model/project_ref.go [hunk 1]
```diff
@@ -2245,54 +2250,78 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
 }
 
 // ValidateProjectCreation returns a boolean if you should surface the error or not.
-func ValidateProjectCreation(projectId string, config *evergreen.Settings, projectRef *ProjectRef) (bool, error) {
+func ValidateProjectCreation(projectId string, config *evergreen.Settings, projectRef *ProjectRef) (int, error) {
```
Comment (malikchaya2) [inline model/project_ref.go:2253]: Can we rename this to include the word `limit`? Like `CheckProjectLimit` or `ValidateProjectCreationLimit`?

File: model/project_ref.go [hunk 2]
```diff
@@ -2245,54 +2250,78 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
-		return true, errors.Wrap(err, "getting number of projects")
+		return http.StatusInternalServerError, errors.Wrapf(err, "getting project '%s'", projectId)
 	}
-	allEnabledProjectsId := []string{}
-	for _, p := range allEnabledProjects {
-		allEnabledProjectsId = append(allEnabledProjectsId, p.Id)
+	if originalRef != nil {
+		if projectRef == nil {
+			// Should be unreachable
+			return http.StatusInternalServerError, errors.New("cannot create project while defaulting to repo")
```
Comment (malikchaya2) [inline model/project_ref.go:2264]: Is this reached if someone tries to copy an existing project that defaults to the repo? In that case should we change the error message to say "create or copy"?

File: model/project_ref.go [hunk 3]
```diff
@@ -2245,54 +2250,78 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
 }
 
 // ValidateProjectCreation returns a boolean if you should surface the error or not.
```
[the remark this answers is not in the document]
Comment (malikchaya2) [reply model/project_ref.go:2252]: +1

**Review by malikchaya2 — APPROVED** [review]
approved with some requests

File: graphql/mutation_resolver.go [hunk 4]
```diff
@@ -408,16 +408,19 @@ func (r *mutationResolver) CreateProject(ctx context.Context, project restModel.
...
-			if apiErr.StatusCode == http.StatusBadRequest {
-				return nil, InputValidationError.Send(ctx, apiErr.Message)
+	if created, err := data.CreateProject(ctx, evergreen.GetEnvironment(), dbProjectRef, u); err != nil {
+		if !created {
+			apiErr, ok := err.(gimlet.ErrorResponse)
+			if ok {
+				if apiErr.StatusCode == http.StatusBadRequest {
+					return nil, InputValidationError.Send(ctx, apiErr.Message)
+				}
+				// StatusNotFound and other error codes are really internal errors bc we determine this input
```
Comment (ablack12) [inline graphql/mutation_resolver.go:418]: Do we need this return statement (isn't it just the same as the one below)?

File: model/project_ref.go [hunk 5]
```diff
@@ -1896,27 +1904,24 @@ func SaveProjectPageForSection(projectId string, p *ProjectRef, section ProjectP
 		if p.TracksPushEvents != nil {
 			setUpdate[ProjectRefTracksPushEventsKey] = p.TracksPushEvents
 		}
+		config, error := evergreen.GetConfig()
```
Comment (ablack12) [inline model/project_ref.go:1907]: shouldn't use error since this is the package name. If you want to have distinct error names I would make this one specific, like configErr. It's pretty unclear to me what's going on here though.

File: model/project_ref.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ablack12) [reply model/project_ref.go:2252]: This also needs some more details about what it's validating, as a public function. And it appears that it can be expected for the inputted projectRef to be nil. Can you expand on that?

File: model/project_ref.go [hunk 6]
```diff
@@ -2245,54 +2250,78 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
 // ValidateProjectCreation returns a boolean if you should surface the error or not.
-func ValidateProjectCreation(projectId string, config *evergreen.Settings, projectRef *ProjectRef) (bool, error) {
+func ValidateProjectCreation(projectId string, config *evergreen.Settings, projectRef *ProjectRef) (int, error) {
 	if config.ProjectCreation.TotalProjectLimit == 0 || config.ProjectCreation.RepoProjectLimit == 0 {
-		return false, nil
+		return 0, nil
 	}
-	catcher := grip.NewBasicCatcher()
-	allEnabledProjects, err := GetEnabledProjects()
+	originalRef, err := findOneProjectRefQ(byId(projectId))
```
Comment (ablack12) [inline model/project_ref.go:2257]: Why not call FindMergedProjectRef if you end up merging with repo anyway? (Otherwise FindBranchProjectRef)

File: model/project_ref.go [hunk 7]
```diff
@@ -2245,54 +2250,78 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
+			return http.StatusInternalServerError, errors.Wrapf(err, "getting merged project for '%s'", originalRef.Id)
+		}
 	}
-	// No need to validate if project was enabled to begin with.
-	if utility.StringSliceContains(allEnabledProjectsId, projectId) {
-		return false, nil
+	var newRef *ProjectRef
+	if projectRef == nil {
+		// Default to repo case
+		repoRef, err := FindOneRepoRef(originalRef.Id)
```
Comment (ablack12) [inline model/project_ref.go:2274]: Should add a comment to explain why this needs repo ref.

File: model/project_ref.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ablack12) [reply model/project_ref.go:2252]: This is also now about enabling a project, not creating one

File: model/project_ref.go [hunk 8]
```diff
@@ -2245,54 +2250,78 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
+		}
+		originalRef, err = GetProjectRefMergedWithRepo(*originalRef)
+		if err != nil {
+			return http.StatusInternalServerError, errors.Wrapf(err, "getting merged project for '%s'", originalRef.Id)
+		}
 	}
-	// No need to validate if project was enabled to begin with.
-	if utility.StringSliceContains(allEnabledProjectsId, projectId) {
-		return false, nil
+	var newRef *ProjectRef
```
Comment (ablack12) [inline model/project_ref.go:2271]: refToValidate might be clearer since we aren't creating a new project

File: model/project_ref.go [hunk 9]
```diff
@@ -2245,54 +2250,78 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
+	if projectRef == nil {
```
Comment (ablack12) [inline model/project_ref.go:2272]: In this case we could instead just set projectRef to ProjectRef{RepoRefId: <id needed>} and then call GetProjectRefMergedWithRepo for both cases

File: model/project_ref.go [hunk 10]
```diff
@@ -2245,54 +2250,78 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
-	pRef, err := findOneProjectRefQ(byId(projectId))
-	if err != nil {
-		return true, errors.Wrapf(err, "getting project '%s", projectId)
-	}
-	if pRef != nil {
-		// Defaulting to repo
-		if projectRef == nil {
-			projectRef, err = GetProjectRefMergedWithRepo(*pRef)
+	catcher := grip.NewBasicCatcher()
+	if originalRef == nil || newRef.IsEnabled() {
```
Comment (ablack12) [inline model/project_ref.go:2293]: These if statements need some rethinking or commenting. What is going on here? Why do we check if originalRef == nil in both if statements? Does that imply "defaultToRepo"? We should set that at the top instead to make it way clearer what's going on.

File: model/project_ref.go (hunk 9 above)
Comment (ablack12) [reply model/project_ref.go:2272]: I'm actually confused about why all of this is necessary. Why can't we just pass in mergedSection https://github.com/evergreen-ci/evergreen/blob/d80b595462048b97a7557e24f57483eecd9beb92/rest/data/project_settings.go#L217 to understand if the thing is going to be enabled or not?

**Review by ablack12 — CHANGES_REQUESTED** [review]
Does this handle the "isRepo" case where a repo is being set to enabled and all the branches are disabled?

I'm thinking this is too complex. Do you know who is using a repo to determine whether a branch is enabled or not? It may be more worth removing that option so we can _way_ simplify this validation. I'm also not fully sure this needs to be as complex as it is though.

File: rest/data/project.go [hunk 11]
```diff
@@ -108,12 +109,19 @@ func CreateProject(ctx context.Context, env evergreen.Environment, projectRef *m
 		}
 	}
 	if err := VerifyUniqueProject(projectRef.Id); err != nil {
-		return err
+		return false, err
 	}
 	// Always warn because created projects are never enabled.
-	_, err := model.ValidateProjectCreation(projectRef.Id, env.Settings(), projectRef)
+	catcher := grip.NewBasicCatcher()
```
Comment (ablack12) [inline rest/data/project.go:115]: Can we call this warningCatcher or something?

File: model/project_ref.go (hunk 9 above)
Comment (bynn) [reply model/project_ref.go]: My reasoning for it was that when we call the validate function, we don't always have the merged ref in other places so I'd just take the minimal info and handle every case (like defaulting to repo) in the function.

But I think it does make more sense to make the function simpler and handle the other cases separately

File: model/project_ref.go (hunk 7 above)
Comment (bynn) [reply model/project_ref.go]: deleted

File: model/project_ref.go (hunk 10 above)
Comment (bynn) [reply model/project_ref.go]: the original == nil checks if the project previously exists or not and that means we're creating/copying

I don't think putting it on the top works because there's an extra condition for checking the owner/repo and that would duplicate more code. I put in comments to make this clearer

File: rest/data/project.go (hunk 11 above)
Comment (bynn) [reply rest/data/project.go]: done

File: model/project_ref.go (hunk 2 above)
Comment (bynn) [reply model/project_ref.go]: removed

File: model/project_ref.go [hunk 12]
```diff
@@ -2245,54 +2250,78 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
 	if err != nil {
-		return true, errors.Wrap(err, "getting number of projects")
+		return http.StatusInternalServerError, errors.Wrapf(err, "getting project '%s'", projectId)
 	}
-	allEnabledProjectsId := []string{}
-	for _, p := range allEnabledProjects {
-		allEnabledProjectsId = append(allEnabledProjectsId, p.Id)
+	if originalRef != nil {
+		if projectRef == nil {
+			// Should be unreachable
```
[the remark this answers is not in the document]
Comment (bynn) [reply model/project_ref.go]: removed

File: model/project_ref.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (bynn) [reply model/project_ref.go]: updated

File: model/project_ref.go (hunk 1 above)
Comment (bynn) [reply model/project_ref.go]: renamed to `ValidateEnabledProjectsLimit`

File: model/project_ref.go [hunk 13]
```diff
@@ -2244,55 +2260,52 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
 	return res, nil
 }
 
-// ValidateProjectCreation returns a boolean if you should surface the error or not.
-func ValidateProjectCreation(projectId string, config *evergreen.Settings, projectRef *ProjectRef) (bool, error) {
+// ValidateEnabledProjectsLimit takes in a merged project ref and validates if the current project is allowed to be enabled.
+// Returns a status code and error we are already at limit with enabled projects.
+func ValidateEnabledProjectsLimit(projectId string, config *evergreen.Settings, mergedRefToValidate ProjectRef) (int, error) {
```
Comment (ablack12) [inline model/project_ref.go:2265]: Preference for mergedRefToValidate to be a pointer since that's better for passing -- we don't need to check if it's nil if that's not correct to pass in.

File: model/project_ref.go [hunk 14]
```diff
@@ -2244,55 +2260,52 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
 	return res, nil
 }
 
-// ValidateProjectCreation returns a boolean if you should surface the error or not.
-func ValidateProjectCreation(projectId string, config *evergreen.Settings, projectRef *ProjectRef) (bool, error) {
+// ValidateEnabledProjectsLimit takes in a merged project ref and validates if the current project is allowed to be enabled.
+// Returns a status code and error we are already at limit with enabled projects.
```
Comment (ablack12) [inline model/project_ref.go:2264]: **if we

File: model/project_ref.go [hunk 15]
```diff
@@ -2244,55 +2260,52 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
+func ValidateEnabledProjectsLimit(projectId string, config *evergreen.Settings, mergedRefToValidate ProjectRef) (int, error) {
 	if config.ProjectCreation.TotalProjectLimit == 0 || config.ProjectCreation.RepoProjectLimit == 0 {
```
Comment (ablack12) [inline model/project_ref.go:2266]: If mergedRefToValidate is disabled then we should return early.

File: model/project_ref.go [hunk 16]
```diff
@@ -2244,55 +2260,52 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
-		return false, nil
+		return 0, nil
```
Comment (ablack12) [inline model/project_ref.go:2267]: Comment mentions this should be a status code so should we return http.StatusOk?

File: model/project_ref.go [hunk 17]
```diff
@@ -2244,55 +2260,52 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
 	}
-	catcher := grip.NewBasicCatcher()
-	allEnabledProjects, err := GetEnabledProjects()
+	originalRef, err := FindMergedProjectRef(projectId, "", false)
```
Comment (ablack12) [inline model/project_ref.go:2269]: If mergedRefToValidate is disabled then we can return before querying the original project.

File: model/project_ref.go [hunk 18]
```diff
@@ -2244,55 +2260,52 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
+		return http.StatusInternalServerError, errors.Wrapf(err, "getting project '%s'", projectId)
 	}
 
-	// Only error if we are enabling a previously disabled project.
-	shouldError := false
-	pRef, err := findOneProjectRefQ(byId(projectId))
-	if err != nil {
-		return true, errors.Wrapf(err, "getting project '%s", projectId)
+	catcher := grip.NewBasicCatcher()
+	if originalRef == nil || !originalRef.IsEnabled() {
```
Comment (ablack12) [inline model/project_ref.go:2275]: originalRef shouldn't be nil right? I'm not sure I understand this. Shouldn't we just error if it's nil?

File: model/project_ref.go [hunk 19]
```diff
@@ -2244,55 +2260,52 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
+		allEnabledProjects, err := GetNumberOfEnabledProjects()
+		if err != nil {
+			return http.StatusInternalServerError, errors.Wrap(err, "getting number of projects")
+		}
+		if originalRef == nil {
+			allEnabledProjects++
+		}
+		// Only validate total project limit if it's a new project or was previously disabled.
```
Comment (ablack12) [inline model/project_ref.go:2283]: Can you explain the "if it's a new project" part? Shouldn't we only validate if it's enabled period?

Comment (ablack12) [reply model/project_ref.go:2283]: This is also a confusing place for this comment since we don't check anything like that in this if statement.

File: model/project_ref.go [hunk 20]
```diff
@@ -2244,55 +2260,52 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
+		if allEnabledProjects >= config.ProjectCreation.TotalProjectLimit {
+			catcher.Errorf("total project limit of %d reached", config.ProjectCreation.TotalProjectLimit)
```
Comment (ablack12) [inline model/project_ref.go:2285]: enabled project

File: rest/data/project.go [hunk 21]
```diff
@@ -108,12 +109,19 @@ func CreateProject(ctx context.Context, env evergreen.Environment, projectRef *m
...
 	// Always warn because created projects are never enabled.
-	_, err := model.ValidateProjectCreation(projectRef.Id, env.Settings(), projectRef)
+	warningCatcher := grip.NewBasicCatcher()
+	statusCode, err := model.ValidateEnabledProjectsLimit(projectRef.Id, env.Settings(), *projectRef)
 	if err != nil {
-		// TODO EVG-18784: Return graphql warning
+		if statusCode != http.StatusBadRequest {
+			return false, gimlet.ErrorResponse{
+				StatusCode: statusCode,
+				Message:    errors.Wrapf(err, "inserting project '%s'", projectRef.Identifier).Error(),
```
Comment (ablack12) [inline rest/data/project.go:121]: this message needs to be updated

File: model/project_ref.go (hunk 18 above)
Comment (ablack12) [reply model/project_ref.go:2275]: oh is this the "created" case so we want to check and warn? Let's think of ways to refactor this to be clearer. We could assign a variable `isNewProject := originalRef == nil` and comment to explain this, and comment in the function header that if it's a new project, we validate as though it were enabled to see if that would be a problem.

Another idea is we could pass in a variable to the top-level that indicates if we're assuming it's enabled.

File: model/project_ref.go [hunk 22]
```diff
@@ -2244,55 +2260,52 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
-	pRef, err := findOneProjectRefQ(byId(projectId))
-	if err != nil {
-		return true, errors.Wrapf(err, "getting project '%s", projectId)
+	catcher := grip.NewBasicCatcher()
+	if originalRef == nil || !originalRef.IsEnabled() {
+		allEnabledProjects, err := GetNumberOfEnabledProjects()
+		if err != nil {
+			return http.StatusInternalServerError, errors.Wrap(err, "getting number of projects")
+		}
+		if originalRef == nil {
```
Comment (ablack12) [inline model/project_ref.go:2280]: I would think we'd want to do this even if original ref is disabled, since the idea is that we should be enabling the given ref.

File: model/project_ref.go [hunk 23]
```diff
@@ -2244,55 +2260,52 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
+			catcher.Errorf("total project limit of %d reached", config.ProjectCreation.TotalProjectLimit)
+		}
 	}
-	if pRef != nil {
-		// Defaulting to repo
-		if projectRef == nil {
-			projectRef, err = GetProjectRefMergedWithRepo(*pRef)
+
+	// Only validate owner/repo limit if it's a new project, was previously disabled, or we're updating owner/repo.
+	if originalRef == nil || !originalRef.IsEnabled() || originalRef.Owner != mergedRefToValidate.Owner || originalRef.Repo != mergedRefToValidate.Repo {
```
Comment (ablack12) [inline model/project_ref.go:2290]: this would be a good spot for a tiny private helper function shouldValidateOwnerRepoLimit or something, and can combine this if with the below

File: rest/data/project_settings.go [hunk 24]
```diff
@@ -256,6 +260,15 @@ func SaveProjectSettingsForSection(ctx context.Context, projectId string, change
 			}
 		}
 
+		config, err := evergreen.GetConfig()
+		if err != nil {
+			return nil, errors.Wrap(err, "getting evergreen config")
+		}
+		_, err = model.ValidateEnabledProjectsLimit(projectId, config, *mergedSection)
```
Comment (ablack12) [inline rest/data/project_settings.go:267]: should validateEnabledProjectsLimit take in the original ref as well so we don't re-query mergedbeforeRef? (for the create case can pass in an empty project ref, or could refactor that function so there are two different functions that reuse a lot of the code)

**Review by ablack12 — CHANGES_REQUESTED** [review]
Left a lot of comments but this is looking better! I think there are a few edge cases we need to add tests for and some room to refactor a bit more.

File: model/project_ref.go (hunk 23 above)
Comment (bynn) [reply model/project_ref.go]: done

File: graphql/mutation_resolver.go [hunk 25]
```diff
@@ -408,16 +408,17 @@ func (r *mutationResolver) CreateProject(ctx context.Context, project restModel.
...
+				if apiErr.StatusCode == http.StatusBadRequest {
+					return nil, InputValidationError.Send(ctx, apiErr.Message)
+				}
 			}
-			// StatusNotFound and other error codes are really internal errors bc we determine this input
-			return nil, InternalServerError.Send(ctx, apiErr.Message)
+			return nil, InternalServerError.Send(ctx, err.Error())
 		}
-		return nil, InternalServerError.Send(ctx, err.Error())
+		graphql.AddError(ctx, PartialError.Send(ctx, err.Error()))
```
Comment (khelif96) [inline graphql/mutation_resolver.go:421]: Can you explain this logic? Why are we adding an error to the response?

Comment (bynn) [reply graphql/mutation_resolver.go:421]: this is from sophies comment 

~		// Use AddError to bypass gqlgen restriction that data and errors cannot be returned in the same response
		// https://github.com/99designs/gqlgen/issues/1191
~

File: model/project_ref.go [hunk 26]
```diff
@@ -2244,55 +2266,57 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
-			projectRef, err = GetProjectRefMergedWithRepo(*pRef)
-			if err != nil {
-				return true, errors.Wrapf(err, "getting merged project for '%s'", projectId)
-			}
+	isNewProject := originalMergedRef == nil
+	catcher := grip.NewBasicCatcher()
+	if shouldValidateTotalProjectLimit(isNewProject, originalMergedRef) {
+		allEnabledProjects, err := GetNumberOfEnabledProjects()
+		if err != nil {
+			return http.StatusInternalServerError, errors.Wrap(err, "getting number of projects")
```
Comment (ablack12) [inline model/project_ref.go:2299]: of enabled projects

File: model/project_ref.go [hunk 27]
```diff
@@ -2244,55 +2266,57 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
+	if shouldValidateOwnerRepoLimit(isNewProject, config, originalMergedRef, mergedRefToValidate) {
+		enabledOwnerRepoProjects, err := GetNumberOfEnabledProjectsForOwnerRepo(mergedRefToValidate.Owner, mergedRefToValidate.Repo)
 		if err != nil {
-			return true, errors.Wrapf(err, "getting number of projects for '%s/%s'", projectRef.Owner, projectRef.Repo)
+			return http.StatusInternalServerError, errors.Wrapf(err, "getting number of projects for '%s/%s'", mergedRefToValidate.Owner, mergedRefToValidate.Repo)
 		}
-		if len(enabledOwnerRepoProjects) >= config.ProjectCreation.RepoProjectLimit {
-			catcher.Errorf("owner repo limit of %d reached for '%s/%s'", config.ProjectCreation.RepoProjectLimit, projectRef.Owner, projectRef.Repo)
+		if enabledOwnerRepoProjects >= config.ProjectCreation.RepoProjectLimit {
+			catcher.Errorf("owner repo limit of %d reached for '%s/%s'", config.ProjectCreation.RepoProjectLimit, mergedRefToValidate.Owner, mergedRefToValidate.Repo)
```
Comment (ablack12) [inline model/project_ref.go:2312]: since this error will go to users, should be a bit clearer: "enabled project limit" might be clearer than "owner repo limit" (we specify owner/repo at the end so that clarifies it i think)

File: model/project_ref.go (hunk 15 above)
Comment (ablack12) [reply model/project_ref.go:2290]: looks like we handle this by only calling the function if the mergedRefToValidate is enabled

Comment (ablack12) [reply model/project_ref.go:2290]: Maybe we can tweak the first comment to something like "and validates project limits, assuming the given project is going to be enabled". Also isn't quite accurate since it says it takes in 1 project ref.

File: model/project_ref.go [hunk 28]
```diff
@@ -2244,55 +2266,57 @@ func (p *ProjectRef) GetGithubProjectConflicts() (GithubProjectConflicts, error)
...
+// - the original project was disabled
+func shouldValidateTotalProjectLimit(isNewProject bool, originalMergedRef *ProjectRef) bool {
+	return isNewProject || !originalMergedRef.IsEnabled()
+}
+
+// shouldValidateOwnerRepoLimit will return true if:
+// - we are creating a new project
+// - the original project was disabled
+// - the owner or repo has changed
+// - the owner/repo is part of exception
```
Comment (ablack12) [inline model/project_ref.go:2280]: if the owner repo is NOT an exception right? Maybe "will return true if the owner/repo is not exempt from validation and: <list>"

**Review by ablack12 — APPROVED** [review]
LGTM with nits

File: model/project_ref.go (hunk 28 above)
Comment (bynn) [reply model/project_ref.go:2280]: yes!

File: model/project_ref.go (hunk 15 above)
Comment (bynn) [reply model/project_ref.go:2290]: done

File: model/project_ref.go (hunk 26 above)
Comment (bynn) [reply model/project_ref.go:2299]: done

File: model/project_ref.go (hunk 27 above)
Comment (bynn) [reply model/project_ref.go:2312]: done
